frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java

```diff
@@ -46,185 +45,217 @@
+
+    LOGGER.logInfo(
+        "select: TaskRole: [%s] Resource: [%s]", taskRoleName, requestResource);
+    String requestNodeLabel = am.getRequestManager().getTaskPlatParams().get(taskRoleName).getTaskNodeLabel();
+    String requestNodeGpuType = am.getRequestManager().getTaskPlatParams().get(taskRoleName).getTaskNodeGpuType();
+    int pendingTaskNumber = am.getStatusManager().getUnAssociatedTaskCount(taskRoleName);
+    List<ValueRange> reUsePorts = null;
+
+    // Prefer to use previous successfully allocated ports. if no successfully ports, try to re-use the "Requesting" ports.
+    if (am.getRequestManager().getTaskRoles().get(taskRoleName).getUseTheSamePorts()) {
+      reUsePorts = am.getStatusManager().getLiveAssociatedContainerPorts(taskRoleName);
+      if (ValueRangeUtils.getValueNumber(reUsePorts) <= 0 && requestedPortsCache.containsKey(taskRoleName)) {
+        reUsePorts = requestedPortsCache.get(taskRoleName);
+        // the cache only guide the next task to use previous requesting port.
+        requestedPortsCache.remove(taskRoleName);
       }
+    }
+    SelectionResult result = select(requestResource, requestNodeLabel, requestNodeGpuType, pendingTaskNumber, reUsePorts);
+    if (!requestedPortsCache.containsKey(taskRoleName) && pendingTaskNumber > 1)
+      requestedPortsCache.put(taskRoleName, result.getOptimizedResource().getPortRanges());
+    return result;
+  }
 
-      LOGGER.logWarning(
-          "select: The Request will be relaxed to RM");
+  @VisibleForTesting
+  public synchronized SelectionResult select(ResourceDescriptor requestResource, String requestNodeLabel, String requestNodeGpuType, int pendingTaskNumber) throws NotAvailableException {
+    return select(requestResource, requestNodeLabel, requestNodeGpuType, pendingTaskNumber, null);
+  }
+
+  public synchronized SelectionResult select(ResourceDescriptor requestResource, String requestNodeLabel, String requestNodeGpuType,
+      int pendingTaskNumber, List<ValueRange> reUsePorts) throws NotAvailableException {
+
+    LOGGER.logInfo(
+        "select: Request: Resource: [%s], NodeLabel: [%s], NodeGpuType: [%s], TaskNumber: [%d], ReUsePorts: [%s]",
+        requestResource, requestNodeLabel, requestNodeGpuType, pendingTaskNumber, ValueRangeUtils.toString(reUsePorts));
+
+    randomizeNodes();
+    filterNodesByNodeLabel(requestNodeLabel);
+    filterNodesByGpuType(requestNodeGpuType);
+    if (!am.getConfiguration().getLauncherConfig().getAmAllowNoneGpuJobOnGpuNode()) {
+      int jobTotalRequestGpu = am.getRequestManager().getTotalGpuCount();
+      filterNodesForNonGpuJob(jobTotalRequestGpu);
+    }
+
+    ResourceDescriptor optimizedRequestResource = YamlUtils.deepCopy(requestResource, ResourceDescriptor.class);
+    if (ValueRangeUtils.getValueNumber(reUsePorts) > 0) {
+      LOGGER.logInfo(
+          "select: re-use pre-allocated ports: [%s]", ValueRangeUtils.toString(reUsePorts));
+      optimizedRequestResource.setPortRanges(reUsePorts);
     }
```

Comment:
Have you handled the user explicitly specify GpuAttribute or specific Ports case?

Reply:
The  specify GpuAttribute or specific Ports  are in the logic. 
if specific, the reusePorts  is the specific one.

---
In reply to: [174750084](https://github.com/Microsoft/pai/pull/318#discussion_r174750084) [](ancestors = 174750084)